src/binary_format.cc

```diff
@@ -0,0 +1,254 @@
+/**
+ * This file contains binary formatting methods for FactorGraphs and
+ * CompactFactorGraphs. We think this file is a good place to put the
+ * definitions of these methods as it conveys the intentions quite clearly
+ * (i.e. for binary formatting of these objects).
+ */
+#include "common.h"
+#include "binary_format.h"
+#include "factor_graph.h"
+#include "factor.h"
+#include "variable.h"
+#include <fstream>
+#include <iostream>
+#include <cstdint>
+
+namespace dd {
+
+// Read meta data file, return Meta struct
+FactorGraphDescriptor read_meta(const std::string &meta_file) {
+  FactorGraphDescriptor meta;
+  std::ifstream file;
+  file.open(meta_file.c_str());
+  std::string buf;
+  getline(file, buf, ',');
+  meta.num_weights = atoll(buf.c_str());
+  getline(file, buf, ',');
+  meta.num_variables = atoll(buf.c_str());
+  getline(file, buf, ',');
+  meta.num_factors = atoll(buf.c_str());
+  getline(file, buf, ',');
+  meta.num_edges = atoll(buf.c_str());
+  file.close();
+  return meta;
+}
+
+void FactorGraph::load_weights(const std::string &filename) {
+  std::ifstream file;
+  file.open(filename, std::ios::in | std::ios::binary);
+
+  long long count = 0;
+  long long id;
+  bool isfixed;
+  char padding;
+  double initial_value;
+  while (file.good()) {
+    // read fields
+    file.read((char *)&id, 8);
+    file.read((char *)&padding, 1);
+    if (!file.read((char *)&initial_value, 8)) break;
+    // convert endian
+    id = be64toh(id);
+    isfixed = padding;
+    long long tmp = be64toh(*(uint64_t *)&initial_value);
+    initial_value = *(double *)&tmp;
+
+    // load into factor graph
+    weights[id] = Weight(id, initial_value, isfixed);
+    ++count;
+  }
+  size.num_weights += count;
+  file.close();
+}
+
+void FactorGraph::load_variables(const std::string &filename) {
+  std::ifstream file;
+  file.open(filename, std::ios::in | std::ios::binary);
+
+  long long count = 0;
+  long long id;
+  char isevidence;
+  double initial_value;
+  short type;
+  long long edge_count;
+  long long cardinality;
+  while (file.good()) {
+    // read fields
+    file.read((char *)&id, 8);
+    file.read((char *)&isevidence, 1);
+    file.read((char *)&initial_value, 8);
+    file.read((char *)&type, 2);
+    file.read((char *)&edge_count, 8);
+    if (!file.read((char *)&cardinality, 8)) break;
+
+    // convert endian
+    id = be64toh(id);
+    type = be16toh(type);
+    long long tmp = be64toh(*(uint64_t *)&initial_value);
+    initial_value = *(double *)&tmp;
+    edge_count = be64toh(edge_count);
+    cardinality = be64toh(cardinality);
+
+    dprintf(
+        "----- id=%lli isevidence=%d initial=%f type=%d edge_count=%lli"
+        "cardinality=%lli\n",
+        id, isevidence, initial_value, type, edge_count, cardinality);
+
+    ++count;
+
+    variable_domain_type_t type_const;
+    switch (type) {
+      case 0:
+        type_const = DTYPE_BOOLEAN;
+        break;
+      case 1:
+        type_const = DTYPE_MULTINOMIAL;
```

Comment:
Shall we rename this to 'categorical'? (boolean::binomial, categorical::multinomial)